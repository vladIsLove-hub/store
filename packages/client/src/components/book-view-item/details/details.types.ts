import { Book } from '../../booklist/BookListTypes';

export enum DetailsItemType {
  PlainTextValue = 'PlainTextValue',
  SingleEntityValue = 'SingleEntityValue',
  MultipleEntityValue = 'MultipleEntityValue',
}

export type PlainTextValue = string | number;

export type SingleEntityValue = {
  id: string;
  name: string;
};

export type MultipleEntityValue = SingleEntityValue[];

export type DetailsItemProps = {
  type: string;
  value: PlainTextValue | SingleEntityValue | MultipleEntityValue;
  title: string;
};

export type DetailsProps = {
  book: Book;
};

export type DetailsItemLabelProps = {
  text: string;
};

export type DetailsItemPrintTextProps = DetailsItemLabelProps;

export type DetailsItemFieldRenderProps = {
  title: string;
  entityValues: string;
  entityList?: MultipleEntityValue;
};
