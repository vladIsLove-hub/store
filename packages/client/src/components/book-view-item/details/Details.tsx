import React from 'react';

import { DetailsProps, DetailsItemType } from './details.types';
import DetailsItem from './DetailsItem';

const Details: React.FC<DetailsProps> = ({ book }) => {
  return (
    <ul data-testid="details" className="book-view__details-list">
      <DetailsItem value={book.authors} title={'Authors'} type={DetailsItemType.MultipleEntityValue} />
      <DetailsItem value={book.category} title={'Category'} type={DetailsItemType.SingleEntityValue} />
      <DetailsItem value={book.publisher} title={'Publisher'} type={DetailsItemType.SingleEntityValue} />
      <DetailsItem value={`${book.ageLimit}+`} title={'Age limit'} type={DetailsItemType.PlainTextValue} />
      <DetailsItem value={book.publicationYear} title={'Publication year'} type={DetailsItemType.PlainTextValue} />
      <DetailsItem value={book.numberOfPages} title={'Pages'} type={DetailsItemType.PlainTextValue} />
      <DetailsItem value={book.format.join(' x ')} title={'Format'} type={DetailsItemType.PlainTextValue} />
    </ul>
  );
};

export default Details;
