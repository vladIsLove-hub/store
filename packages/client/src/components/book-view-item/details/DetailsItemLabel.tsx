import React from 'react';

import { DetailsItemLabelProps } from './details.types';

const DetailsItemLabel: React.FC<DetailsItemLabelProps> = ({ text }) => {
  return (
    <div className="book-view__details-list-item-left">
      <span className="book-view__details-list-item-name">{text}</span>
    </div>
  );
};

export default DetailsItemLabel;
