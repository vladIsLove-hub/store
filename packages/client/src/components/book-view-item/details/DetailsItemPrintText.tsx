import React from 'react';

import { DetailsItemPrintTextProps } from './details.types';

const DetailsItemPrintText: React.FC<DetailsItemPrintTextProps> = ({ text }) => {
  return (
    <div className="book-view__details-list-item-right">
      <span data-testid="item-value" title={text} className="book-view__details-list-item-value">
        {text}
      </span>
    </div>
  );
};

export default DetailsItemPrintText;
