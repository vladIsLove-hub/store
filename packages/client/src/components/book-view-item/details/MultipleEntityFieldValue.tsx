import React from 'react';
import Accordion from 'react-bootstrap/Accordion';

import { DetailsItemFieldRenderProps } from './details.types';
import DetailsItemPrintText from './DetailsItemPrintText';
import DetailsItemLabel from './DetailsItemLabel';

const MultipleEntityValue: React.FC<DetailsItemFieldRenderProps> = ({ title, entityValues, entityList }) => {
  return (
    <>
      <Accordion className="book-view_details-authors-menu">
        <Accordion.Item eventKey="0">
          <Accordion.Header>Authors: </Accordion.Header>
          <Accordion.Body>
            {entityList && (
              <ul>
                {entityList.map(entity => {
                  return <li key={entity.id}>{entity.name}</li>;
                })}{' '}
              </ul>
            )}
          </Accordion.Body>
        </Accordion.Item>
      </Accordion>
      <li className="book-view__details-list-item book-view__details-authors-info">
        <DetailsItemLabel text={title} />
        <DetailsItemPrintText text={entityValues} />
      </li>
    </>
  );
};

export default MultipleEntityValue;
