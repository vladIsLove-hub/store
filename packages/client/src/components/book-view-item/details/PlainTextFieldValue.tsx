import React from 'react';

import { DetailsItemFieldRenderProps } from './details.types';
import DetailsItemPrintText from './DetailsItemPrintText';
import DetailsItemLabel from './DetailsItemLabel';

const PlainTextFieldValue: React.FC<DetailsItemFieldRenderProps> = ({ title, entityValues }) => {
  return (
    <li className="book-view__details-list-item">
      <DetailsItemLabel text={title} />
      <DetailsItemPrintText text={entityValues} />
    </li>
  );
};

export default PlainTextFieldValue;
