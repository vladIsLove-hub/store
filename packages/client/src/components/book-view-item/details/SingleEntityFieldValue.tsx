import React from 'react';

import { DetailsItemFieldRenderProps } from './details.types';
import DetailsItemPrintText from './DetailsItemPrintText';
import DetailsItemLabel from './DetailsItemLabel';

const SingleEntityFieldValue: React.FC<DetailsItemFieldRenderProps> = ({ title, entityValues }) => {
  return (
    <li className="book-view__details-list-item">
      <DetailsItemLabel text={title} />
      <DetailsItemPrintText text={entityValues} />
    </li>
  );
};

export default SingleEntityFieldValue;
