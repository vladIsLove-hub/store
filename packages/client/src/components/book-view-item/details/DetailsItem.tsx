import React from 'react';

import getEntityValues from '../../../utils/getEntityValues';

import { DetailsItemProps, DetailsItemType } from './details.types';
import SingleEntityFieldValue from './SingleEntityFieldValue';
import PlainTextFieldValue from './PlainTextFieldValue';
import MultipleEntityFieldValue from './MultipleEntityFieldValue';

const DetailsItem: React.FC<DetailsItemProps> = ({ type, value, title }) => {
  const { entity, entityValue, entityList, entityNames } = getEntityValues(value);
  switch (type) {
    case DetailsItemType.SingleEntityValue:
      return <SingleEntityFieldValue title={title} entityValues={entity ? entity.name : ''} />;
    case DetailsItemType.MultipleEntityValue:
      return (
        <MultipleEntityFieldValue title={title} entityValues={entityNames ? entityNames : ''} entityList={entityList} />
      );
    default:
      return <PlainTextFieldValue title={title} entityValues={entityValue ? entityValue : ''} />;
  }
};

export default DetailsItem;
