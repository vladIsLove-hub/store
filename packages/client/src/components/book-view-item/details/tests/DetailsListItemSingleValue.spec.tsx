import React from 'react';
import { render, screen } from '@testing-library/react';

import DetailsItem from '../DetailsItem';
import { DetailsItemType } from '../details.types';

describe('DetailsItemSingleValue', () => {
  const title = 'Title';
  const singleValue = {
    id: '1',
    name: 'Name',
  };

  it('should render currently', () => {
    render(<DetailsItem title={title} value={singleValue} type={DetailsItemType.SingleEntityValue} />);
    expect(screen.getByText('Title')).toBeInTheDocument();
    expect(screen.getByText('Name')).toBeInTheDocument();
  });

  it('DetailsItem snapshot for SingleEntityValue data', () => {
    const container = render(
      <DetailsItem title={title} value={singleValue} type={DetailsItemType.SingleEntityValue} />,
    );
    expect(container).toMatchSnapshot();
  });
});
