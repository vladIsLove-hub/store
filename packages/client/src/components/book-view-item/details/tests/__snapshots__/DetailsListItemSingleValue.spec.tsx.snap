// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsItemSingleValue DetailsItem snapshot for SingleEntityValue data 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <li
        class="book-view__details-list-item"
      >
        <div
          class="book-view__details-list-item-left"
        >
          <span
            class="book-view__details-list-item-name"
          >
            Title
          </span>
        </div>
        <div
          class="book-view__details-list-item-right"
        >
          <span
            class="book-view__details-list-item-value"
            data-testid="item-value"
            title="Name"
          >
            Name
          </span>
        </div>
      </li>
    </div>
  </body>,
  "container": <div>
    <li
      class="book-view__details-list-item"
    >
      <div
        class="book-view__details-list-item-left"
      >
        <span
          class="book-view__details-list-item-name"
        >
          Title
        </span>
      </div>
      <div
        class="book-view__details-list-item-right"
      >
        <span
          class="book-view__details-list-item-value"
          data-testid="item-value"
          title="Name"
        >
          Name
        </span>
      </div>
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
